### ISSUE TITLE
PROTECT bug in C code

### ISSUE THREAD

**Issue by magic-lantern** [CUD2V/pccc#37]
Got this from CRAN maintainers:

There is a PROTECT bug in your package pccc (version 1.0.2). The bug manifests itself as "heap-use-after-free" with ASAN (see Additional issues in CRAN checks). One can provoke the problem also with gctorture (e.g. R_GCTORTURE=100) and using this small example which will segfault:

`library(pccc); get_codes(10)`

The problem is in get_codes.cpp, in calls like

     Rcpp::List dx_fixed = Rcpp::List::create(
          Rcpp::wrap(cds.get_dx_fixed_neuromusc()),
          Rcpp::wrap(cds.get_dx_fixed_cvd()),
          Rcpp::wrap(cds.get_dx_fixed_respiratory()),
          Rcpp::CharacterVector::create(),
          Rcpp::CharacterVector::create(), ...

Rcpp::wrap returns an (unprotected) newly allocated SEXP. All uses of Rcpp::wrap inside a call to Rcpp::List (many instances in get_codes.cpp) need to be protected via Rcpp::Shield<SEXP>(), otherwise these get destroyed by allocations when evaluating the other arguments - as in the attached patch.

**Comment by magic-lantern** [comment]
Although all standard CRAN checks pass, some of the ASAN checks have issues. See:

https://www.stats.ox.ac.uk/pub/bdr/memtests/clang-ASAN/pccc/tests/testthat.Rout.fail

And

https://www.stats.ox.ac.uk/pub/bdr/memtests/gcc-ASAN/pccc/tests/testthat.Rout.fail

**Comment by magic-lantern** [comment]
Running:

```r
devtools::test()
```

I discovered a few warnings due to depredations and changes in testthat. 2 of the tests were skipped and 1 returned a warning.

All of these checks passed (well, both win-builder and rhub had a few broken environments but worked on re-submit)

```r
devtools::check_win_devel()
devtools::check_win_release()
devtools::check_win_oldrelease()
devtools::check_rhub()
```

Submitted to CRAN with `devtools::release()`

When (if?) update accepted, will need to add tag in github.

**Comment by magic-lantern** [comment]
Issue resolved, new version (1.0.3) pushed to CRAN and accepted.
